app/tests/test_sms.py

```diff
@@ -209,3 +210,43 @@ def test_unsubscribe(self):
         self.assertEqual("97204", client.zipcode.zipcode)
         self.assertEqual(self.timestamp, client.alerts_disabled_at)
         self.assert_event(client.id, EventType.UNSUBSCRIBE, zipcode="97204")
+
+    def test_feedback(self):
+        # Give feedback before feedback begin command
+        response = self.client.post(
+            "/sms", data={"Body": "Blah Blah Blah", "From": "+13333333333"}
+        )
+        client_id = Client.query.filter_by(identifier="+13333333333").first().id
+
+        self.assertEqual(200, response.status_code)
+        self.assertEqual(1, Client.query.count())
+        self.assertEqual(0, Event.query.count())
+        self.assertEqual(0, Request.query.get_total_count())
+        self.assert_twilio_response(
+            'Unrecognized option "Blah Blah Blah". Reply with M for the menu.',
+            response.data,
+        )
+
+        # Go through feedback flow
+        response = self.client.post("/sms", data={"Body": "4", "From": "+13333333333"})
+        self.assertEqual(200, response.status_code)
+        self.assertEqual(1, Client.query.count())
+        self.assertEqual(1, Event.query.count())
+        self.assertEqual(0, Request.query.get_total_count())
+        self.assert_twilio_response(
+            "Please enter your feedback below:",
+            response.data,
+        )
+        self.assert_event(client_id, EventType.FEEDBACK_BEGIN)
+
+        response = self.client.post(
+            "/sms", data={"Body": "Blah Blah Blah", "From": "+13333333333"}
+        )
+        self.assertEqual(200, response.status_code)
+        self.assertEqual(1, Client.query.count())
+        self.assertEqual(2, Event.query.count())
+        self.assertEqual(0, Request.query.get_total_count())
+        self.assert_twilio_response("Thank you for your feedback!", response.data)
+        self.assert_event(
```

Comment:
Might be worth adding a case to assert what happens if you try to send another random message. It should not say "Thank you for your feedback!" again.

Reply:
Ok, and how can I test the timedelta thing? Like if they try to give feedback more than an hour later.

Reply:
Try clock.advance(60 * 60 + 1)